Extract swift secret files outside of repo

mokagio opened the pull request:
Follow up on _PR_ to extract the remaining secrets outside of the repository. These secrets are `.swift` file containing values that the app needs at build time.

I built on top of the already existing "Run Script" Build Phase, with the following modifications:

- Moved all the code to a dedicated script to make it easier to modify and review
- The automation now looks for the secrets in the new `~/.configure/wordpress-ios/secrets/` folder
- If secrets cannot be found in a Debug build, fall back to the example ones. In a Release build, fail the build early.

~~_Note that this PR is based on top of `move-fastlane-secrets-outside-repository` for ease of review. Ideally, we should land this into 17.7 (see milestone) to test it end-to-end in CI, but it's okay if we don't get there._~~

## How to test

1. Checkout this branch and run a build in one, some, or all of the "WordPress", "WordPress-Alpha", "WordPress-Internal", and "Jetpack" schemes.
2. Verify it shows the example secrets warning:

<img width="1920" alt="Screen Shot 2021-07-05 at 12 10 52 pm" src="https://user-images.githubusercontent.com/1218433/124410561-1b6aa300-dd8e-11eb-9d1d-8aacc4650a5b.png">

3. Try to archive the project in the same scheme
4. Verify the build fails with the error from the script:

<img width="1083" alt="Screen Shot 2021-07-05 at 12 15 29 pm" src="https://user-images.githubusercontent.com/1218433/124410568-1f96c080-dd8e-11eb-80c1-ba1a971e1dbc.png">

5. Run `bundle exec fastlane run configure_apply`
6. Repeat the tests above and verify they succeed

## Regression Notes
1. Potential unintended areas of impact

All the steps above represent a build from Xcode. Interestingly, I found that an extra step was required for the automated build via Fastlane (see inline comments). It's possible there are other steps that might not work in an automated build which I'm not aware of.


2. What I did to test those areas of impact (or what existing automated tests I relied on)

I tested the build locally and ensured it worked. I also run the installable build step for both WordPress and Jetpack and saw them both [succeed](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-iOS/23471/workflows/a4c75067-a5c4-41e0-8186-4fb27c8ff37b).

3. What automated tests I added (or what prevented me from doing so)
N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Extract secrets root definition in ApplyConfiguration.sh build phase

mokagio committed:
Remove file list and input params from secrets build phase

They were unnecessary. Those values are translated in variables that the
build phase can read, but never did. The script defines its inputs and
outsides internally and that's appropriate because it needs logic to
decide which of the secrets to copy. If the script was selecting the
file based on an index in the array of inputs, it would be brittle
because the order might change without the script knowing about it.

Case in point. The file list didn't include the Jetpack secrets, but
the Jetpack target worked nevertheless.

On the other hand, it appears that the output list IS necessary. Without
it, I couldn't get the app to build with the Fastlane automation.

See:

- Build failing at
  https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-iOS/21768/workflows/b9ccaed2-4243-4491-9d59-c1ac5c606c66/jobs/49731
- Build passing* after reintroducing the output list with the file
  https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-iOS/21769/workflows/4cae0dbf-a99c-41a2-b01d-2610251a524f/jobs/49733

* It doesn't really pass, but it does fail in the expected way: with
TestFlight rejecting a build that was successfully uploaded but had an
already submitted build number.

mokagio committed:
Fix bug in mkdir call in ApplyConfiguration.sh build phase

mokagio committed:
Add a `set -e` call to ApplyConfiguration.sh script

So we'll notice the kind of issue addressed in the previous commit
earlier.

mokagio committed:
Move secret Swift files outside of the repository

mokagio committed:
Fail Release builds early if secrets can't be found

mokagio committed:
Remove run script inputs from Jetpack target

mokagio commented on Scripts/BuildPhases/ApplyConfiguration.sh:
```diff
@@ -1,16 +1,21 @@
 # The Secrets file destination
 SECRETS_DESTINATION_FILE="${BUILD_DIR}/Secrets/Secrets.swift"
-mkdir -p "basename $SECRETS_DESTINATION_FILE"
+mkdir -p $(dirname $SECRETS_DESTINATION_FILE)
```
This code never worked but we never noticed because:

1. There was no `set -e` in the script
2. The folder was created anyway because of the "Output Files" parameter in the Run Script configuration

mokagio commented on Scripts/BuildPhases/ApplyConfiguration.sh:
```diff
@@ -47,6 +52,19 @@ if [ -f "$LOCAL_SECRETS_FILE" ]; then
+    ;;
+  *)
+    echo "Applying Example Secrets"
+    echo "warning: Could not find secrets at $SECRETS_ROOT, falling back to $EXAMPLE_SECRETS_FILE. In a release build, this would be an error."
```
Keen on suggestions on the wording of these two errors. The PR description shows how they appear in Xcode. I'd like them to be actionable for new A12s and clear for external contributors. 

When we'll get down to zero warnings, this one will be visible to everyone running the project without secrets. Currently, it gets lost in the noise. I wonder if an emoji could help make it more noticeable.

AliSoftware replied:
What about:
```
echo "error: Unable to find secrets for file ${SECRETS_DESTINATION_FILE} in ${SECRETS_ROOT}. Cannot continue release build. If you are an Automattician, run fastlane configure_apply to update your secrets."
```
and
```
echo "warning: Unable to find secrets for file ${SECRETS_DESTINATION_FILE} in ${SECRETS_ROOT}. Falling back to ${EXAMPLE_SECRETS_FILE} example file. If you are an Automattician, you will need to fix this if you want to be able to do Release builds."
```

A bit verbose maybe, but better be actionable and clear than succinct and unclear 😉

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -16293,10 +16291,8 @@
-				"$(SRCROOT)/Credentials/secrets-manifest.xcfilelist",
 			);
 			inputPaths = (
-				"$(SRCROOT)/../Scripts/BuildPhases/ApplyConfiguration.sh",
```
Interestingly, I had to keep the output paths parameter, otherwise the build would fail, but only when run through `xcodebuild`, not from Xcode.

![image](https://user-images.githubusercontent.com/1218433/124550597-3c0d2880-de74-11eb-8df4-6f6cab8c7589.png)

```
❌  error: Build input file cannot be found: '/Users/gio/Developer/a8c/wpios/DerivedData/Build/Intermediates.noindex/ArchiveIntermediates/Jetpack/BuildProductsPath/Secrets/Secrets.swift' (in target 'Jetpack' from project 'WordPress')
```

AliSoftware replied:
Yep, that's how outputPaths works and what it's for.

I think the fact that it didn't fail from Xcode GUI is a bug/glitch, maybe because the project was already build from your `xcodebuild` test and didn't have to incremental-build when you tried in Xcode?

But the goal is to tell Xcode which files this script phase generates, so that it can resolve the dependency graph when figuring out which targets, dependencies and phases need to be run and which can be skipped in incremental builds if their output are already up-to-date.

It also allows Xcode to know that this file – which is one of the input of the "Compile Sources" build phase, so which needs to be present before that phase runs – will be generated by this script. Since Xcode does the full dependency graph analysis of the various build phases first, and only runs them once it has resolved the whole graph (both in order to avoid starting to build stuff only to fail later if it can predict earlier on that the failure will happen, and to optimize build time in incremental builds by determining what can be skipped), you need to fill in that Output Files here to let it resolve the dependency graph and optimize incremental builds properly.

mokagio committed:
Handle spaces in path in build phase to copy swift secrets

Spaces in paths... _Shakes fists in the air_

jkmassel commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -15658,10 +15658,8 @@
-				"$(SRCROOT)/Credentials/secrets-manifest.xcfilelist",
 			);
 			inputPaths = (
-				"$(SRCROOT)/../Scripts/BuildPhases/ApplyConfiguration.sh",
```
I think we still want to re-run the script if it fails, no?

AliSoftware replied:
Yep, or if its source changed.

I agree we should restore this, and same for line 16299 below.

jkmassel commented on WordPress/Credentials/secrets-manifest.xcfilelist:
```diff
@@ -1,6 +0,0 @@
-#  secrets-manifest.xcfilelist
-
-$(SRCROOT)/../.configure-files/Secrets.swift
```
Should we keep this file and reference the files using `${HOME}/.configure/...` in order to ensure we rebuild if they change?

mokagio replied:
> Should we keep this file

Following up to my comment [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/16806#issuecomment-876000701) I think the answer is yes. Thanks noticing this.

> reference the files using `${HOME}/.configure/...`

That would be great. Xcode has [some issues](https://github.com/mokagio/HomeFolderTest) expanding `$HOME` when used in the project file, but maybe it can manage in a build phase. I'll try 🤞

mokagio replied:
BTW, `$HOME` in the file list file worked 🎉 See f8531a8b809d64cd294111071dd3aa863041820f

jkmassel reviewed (commented):
Left some suggestions for your consideration

mokagio commented:
@jkmassel thank you for your review.

It turns out my understanding of what the input and output parameters are for was flawed. I did some reading:

- https://developer.apple.com/videos/play/wwdc2018/408/
- https://developer.apple.com/documentation/xcode-release-notes/build-system-release-notes-for-xcode-10
- https://nathanwong.co.uk/post/xcode-buildphases/
- https://indiestack.com/2014/12/speeding-up-custom-script-phases/

I have a better picture now and I'll update accordingly. In short, I'll add back and see if I can optimize the way we specify both inputs and outputs.

**Thanks!**

AliSoftware commented on Scripts/BuildPhases/ApplyConfiguration.sh:
```diff
@@ -1,16 +1,21 @@
 #!/bin/sh
 
+set -e
+
```
What about the more common way of setting this on the shebang directly, and add other common options for super-safe mode?

```suggestion
#!/bin/sh -euo pipefail
```

mokagio replied:
I had never seen this, but I like it.

I'm curious to know what makes you say that it is more common? That's not to challenge your statement, but since I'd never seen that syntax, I'm curious to know which authoritative resources I can look into next time 😄 Thanks.

AliSoftware replied:
Ahah i don't have any authoritative source on that, it's just that my experience is opposite to you, as I saw the flags set in the shebang way more often that seeing them set using `set` via a separate command, so I assumed the former was more common 😛

I've often seen, in articles suggesting to set those flags in eg Script Build Phases — to ensure the Xcode build fails as soon as one command in the script fails instead of only looking at exit code of the last command of the script — suggestions to set it on the "Shell" field (eg `/bin/sh -e`) too, so that's what I got used to do… but YMMV 🙃

mokagio replied:
Thanks for the extra context. I like this setup and I'm keen to use it. 

I tried researching whether it was a appropriate but didn't get far, too much conflicting information.

I say let's adopt it (it's neat to have them all in a oneliner) and see how it plays out for us.

jkmassel replied:
Well this thread answers _that_ question for me 😜

AliSoftware commented on Scripts/BuildPhases/ApplyConfiguration.sh:
```diff
@@ -1,16 +1,21 @@
+mkdir -p $(dirname "$SECRETS_DESTINATION_FILE")
 
 # If the WordPress Production Secrets are available for WordPress, use them
 if [ -f "$PRODUCTION_SECRETS_FILE" ] && [ "$BUILD_SCHEME" == "WordPress" ]; then
```
I noticed that in the following lines where we invoke `cp` down below, the input paths to `cp` are not quoted, could be worth adding quotes for safety and future-proofing this.

For example:
```
cp -v "${PRODUCTION_SECRETS_FILE}" "${SECRETS_DESTINATION_FILE}"
```

mokagio replied:
👍

mokagio replied:
Resolved in a43b65b

AliSoftware commented on Scripts/BuildPhases/ApplyConfiguration.sh:
```diff
@@ -1,16 +1,21 @@
 EXAMPLE_SECRETS_FILE="${SRCROOT}/Credentials/Secrets-example.swift"
 
 # The Secrets file destination
 SECRETS_DESTINATION_FILE="${BUILD_DIR}/Secrets/Secrets.swift"
```
I wonder if it would not be better and more idiomatic to put that file in `$(DERIVED_FILE_DIR)/Secrets.swift` instead (and adjust the project references accordingly), since `DERIVED_FILE_DIR` is the location in which Xcode puts files generated during the build (e.g. CoreData generated classes, Intents generated class files, …) and where it expects us to put these kind of generated files (which it then automatically picks to be compiled)…

Maybe not a big deal as long as we are super-sure that whatever `SECRETS_DESTINATION` we use is `.gitignore`'d and does not risk to be commited, just figured I'd mention what the usual conventions are there 🙂

mokagio replied:
`DERIVED_FILE_DIR` is indeed the location where Xcode expects / suggests to put output files, as one can verify by adding a new file to the list:

<img alt="image" src="https://user-images.githubusercontent.com/1218433/127935105-7a9450c3-0101-49c7-a818-22d8705d7057.png" width=300/>

I didn't do it originally because I didn't know about it 😅 . But, after spending some time trying to make it work, I think it wasn't done like that because of how the project is setup. Namely because developers can read `Secrets.swift` via the file navigator to edit it, which I think is a capability we want to keep for external contributors?

By putting the file in the `BUILD_PRODUCTS` we can reference it in the project using the "Relative to Build Products" location

<img alt="image" src="https://user-images.githubusercontent.com/1218433/127935392-0a458627-2497-485a-a867-6cdc000da5e9.png" width=300/>

Unfortunately, there is not "Relative to Derived Sources" option:

<img alt="image" src="https://user-images.githubusercontent.com/1218433/127935512-a7928f63-eb2b-4de5-8eee-4704c8630312.png" width=300/>

(I'm calling it "Derived Sources" because that's one of the [aliases for `DERIVED_FILE_DIR`](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW43))

It is possible to reference the file in the project by adding a custom location in Preferences > Locations > Custom Paths then update the location for the file:

![image](https://user-images.githubusercontent.com/1218433/127935856-e2dea0d7-5cac-45ee-a8ec-9113c5c5a383.png)

With that, the app compiles but the file is still not accessible, it seems like the path used at build time and the one Xcode computes are different 🤔 

<img width="1657" alt="Screen Shot 2021-08-03 at 10 26 06 am" src="https://user-images.githubusercontent.com/1218433/127939491-8654e741-4c9d-4840-bd3c-1bf41104ad7f.png">

Even if we could make it work, the problem I see with that option is that it would require every developer to make that change in their Xcode setup, and repeat it on every new machine. That sounds like it would generate a lot of support requests. Granted, I'm sure we could find a way to automate it as part of the project bootstrap process, but I wonder if it's worth it.

I also tried bypassing the problem by removing the file reference, but that obviously didn't work because then Xcode didn't know where to find the values defined in `Secrets.swift`.

---

**TL;DR** I wasn't able to make it work in a portable way using `DERIVED_FILE_DIR`. I'm keen on suggestions and improvements because the flow is a bit awkward –even the fact of having the example file tracked in the project is a bit confusing– but I think it's work that might deserve dedicated focus later on.

AliSoftware replied:
Agreed with all that, good points. Let's keep it the way it is then

AliSoftware reviewed (commented):
I agree with Jeremy's feedback and suggestion to restore the input files.

I also spotted another issue in the script (which I couldn't inline-comment on since it was not part of this diff and already present before this PR) which is worth fixing: on line 35 of the script where you're about to copy the file for Alpha, the `-f` test is on `INTERNAL_SECRETS_FILE` instead of testing the `ALPHA_SECRETS_FILE` that you're about to copy.

---

Finally, maybe not for this PR and an optimization only for later, but it could be nice to **configure this script as a dedicated "Aggregate Target"** that the WordPress target would depend on, rather than a build phase in the target.

The advantages of doing this are:
 * You can DRY the logic and reuse it between WP and JP (and other targets if needed) without having to repeat the phase in each target
 * But more importantly, Xcode will resolve the dependencies first and will thus execute this Aggregate Target… before trying to build any dependency of the WordPress or Jetpack target, which means that e.g. when you try to Archive, it will error right from the beginning – instead of waiting your time building the dependencies / pods first and only then running the app target afterwards and error-ing during the script build phase only then.

FYI, I gave it a quick try to compare:
<details><summary>12s for clean Archive to fail with the "Aggregate Target as a dependency" technique</summary>

Since Xcode only prepares the directories for other dependant targets but doesn't start building those targets before it runs the first dependency that is my "Generate Secrets" aggregate target, allowing it to fail early.

![image](https://user-images.githubusercontent.com/216089/124916595-2114f280-dff3-11eb-8690-1396a8791fd2.png)
</details>
<details><summary>Compared to 2'47" (!) for clean Archive to fail with the current "Build Phase" technique</summary>

Here Xcode first builds all the dependant targets (i.e. Pods) first, and only then gets to running the WordPress target and thus the build phase, taking way longer to reach the failure point.

![image](https://user-images.githubusercontent.com/216089/124917202-d3e55080-dff3-11eb-97ea-c790616f77af.png)

</summary>

mokagio committed:
Add check to ensure secrets are in input file list

mokagio committed:
Restore secrets input file list

See conversation in
https://github.com/wordpress-mobile/WordPress-iOS/pull/16806

Thanks @jkmassel @AliSoftware

This commit also adds the input to the Jetpack target.

mokagio committed:
Check for alpha secrets file when build configuration is alpha

Before, we were checking internal 😱

Hat tip @AliSoftware.

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/16806#pullrequestreview-701907908

mokagio committed:
Set executable options in Shebang for ApplyConfiguration script

Hat tip @AliSoftware.

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/16806/files?file-filters%5B%5D=.pbxproj&file-filters%5B%5D=.sh&file-filters%5B%5D=.xcfilelist#r666077522

mokagio committed:
Update missing secrets error message to be clearer

At least, the _intent_ is to be clearer. This new wording is more
verbose but should provide better instructions to both internal and
external contributors.

See discussion here:
https://github.com/wordpress-mobile/WordPress-iOS/pull/16806/files?file-filters%5B%5D=.pbxproj&file-filters%5B%5D=.sh&file-filters%5B%5D=.xcfilelist#r666083298

mokagio committed:
Quote all paths used in the `cp` calls in the secrets script

Hat tip @AliSoftware

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/16806/files?file-filters%5B%5D=.pbxproj&file-filters%5B%5D=.sh&file-filters%5B%5D=.xcfilelist#r666079239

mokagio commented:
> Finally, maybe not for this PR and an optimization only for later, but it could be nice to configure this script as a dedicated "Aggregate Target" that the WordPress target would depend on, rather than a build phase in the target.

💯 – This is how Simplenote and WooCommerce iOS do it, so it would also make the setup consistent, on top all the advantages you listed which I agree with. It didn't occur to me to list it as a possible improvement, but we should definitely do it as a followup to this secrets work.

mokagio committed:
Use bash instead of sh in the secrets build phase shebang

We agreed that bash should be our default because it has more features
than plain sh. Unfortunately, I couldn't find the place where we had
that discussion to link back in this message. Sorry.

mokagio committed:
Use a named variable instead of $1 in secrets build phase script

This makes the code slightly more readable

mokagio committed:
Add secrets script using xcfilelist to xcfilelist

mokagio commented:
@AliSoftware @jkmassel I think I addressed all your comments and concerns. Thank you so much for the guidance and patience.

This PR took a long time to finish (assuming you'll approve it) but at least I learned a lot and it made working on https://github.com/Automattic/simplenote-macos/pull/1012 and https://github.com/woocommerce/woocommerce-ios/pull/4713 a breeze.

Could you have another look? Thanks!

jkmassel reviewed (approved):
This LGTM – I was a little confused because I couldn't get the warning to trigger – then I realized it's because I've reviewed this PR previously 🤦 

:shipit:
